cli/src/commands/git/fetch.rs

```diff
@@ -98,8 +106,24 @@ pub fn cmd_git_fetch(
             GitFetchError::InternalGitError(err) => map_git_error(err),
             _ => user_error(err),
         })?;
+        for pattern in &args.branch {
+            if branch_in_remote(&view, remote, pattern, &stats.import_stats) {
+                found_branches.insert(pattern.as_str());
+                continue;
+            }
+        }
         print_git_import_stats(ui, tx.repo(), &stats.import_stats, true)?;
     }
+    for branch in args
+        .branch
+        .iter()
+        .filter(|b| !found_branches.contains(b.as_str()))
+    {
```

Comment:
Perhaps, we don't have to clone `repo.view()` (which is expensive), build `found_branches` set, etc.

Here we can access to `tx.repo()` (= new state) and `tx.base_repo()` (= old state), and we'll need to emit warning if the branch pattern doesn't match either of these repos? Something like (untested):
```rust
let remote_patterns = remotes.iter().map(StringPattern::exact).collect_vec();
for branch in &args.branch {
    // or [views..].iter().cartesian_product(&remote_patterns) ?
    let matched = remote_patterns.iter().any(|remote| tx.repo().view().remote_bookmarks_matching(branch, remote).next().is_some()) || ...;
    if !matched {
        warn ...
    }
}
```

Reply:
Done. This is simpler and I think better.

It would have been nice to issue the warning before the fetch is done, instead of after, but life can't be perfect.

Thanks for the code pointers. Much appreciated as usual.